### PR TITLE
new UI for ai_settings yaml to support multiple configurations

### PR DESCRIPTION
**Pull request by bassie661**
### Summary
Changed the prompt_user flow so the user is now able to configure multiple agents and have their settings stored in ai_settings.yaml. Which now stores as dictionary to allow selecting from multiple key's (configs), save, change or delete them.

When configurations are found in ai_settings.yaml, the prompt starts-up with the following menu:

menu

### Change
- Support storing multiple AI configurations
- Support choose existing AI configuration(s)
- Support create new AI configuration(s)
- Support delete existing AI configuration(s)
- Support change existing AI configuration(s)
   -> e.g. ai_name, ai_role, ai_goals[] and api_budget
- Support for plugins per AI configuration (add, delete, change), plugins are expected to land with #4089 

** Initial plan was to also add support for setting temperature (ai_temp) per AI configuration (add, delete, change) but this clearly needs extra changes, so will create a separate PR for that if this PR gets merged.
** I had some local issues with my git setup, so I had to create a new branch.
** This is a new approach rendered #3711 obsolete and therefor it's closed.

### Test Plan
Added pytests to support future testing.
- test_ai_config.py
- integration\test_setup.py

Although I tested the new functionality thoroughly, it's quite a big change so please help to make sure it's reviewed well.

### PR Quality Checklist
 My pull request is atomic and focuses on a single change.
 I have thoroughly tested my changes with multiple different prompts.
 I have considered potential risks and mitigations for my changes.
 I have documented my changes clearly and comprehensively.
 I have not snuck in any "extra" small tweaks changes.
 I have run the following commands against my code to ensure it passes our linters:
black .
isort .
mypy
autoflake --remove-all-unused-imports --recursive --ignore-init-module-imports autogpt tests --in-place

### CODE REVIEW

File: autogpt/config/ai_config.py [hunk 1]
```diff
@@ -29,14 +28,18 @@ class AIConfig:
         ai_role (str): The description of the AI's role.
         ai_goals (list): The list of objectives the AI is supposed to complete.
         api_budget (float): The maximum dollar value for API calls (0.0 means infinite)
+        plugins (list): The list of objectives the AI is supposed to complete.
     """
 
+    SAVE_FILE = str(Path(os.getcwd()) / "ai_settings.yaml")
```
Comment (erik-megarad) [inline autogpt/config/ai_config.py:34]: Can we guarantee that `cwd` will be the correct dir? Seems like we can use the configured workspace dir.

File: autogpt/config/ai_config.py [hunk 2]
```diff
@@ -29,14 +28,18 @@ class AIConfig:
...
+
     def __init__(
         self,
         ai_name: str = "",
         ai_role: str = "",
         ai_goals: list | None = None,
         api_budget: float = 0.0,
+        plugins: list | None = None,
```
Comment (erik-megarad) [inline autogpt/config/ai_config.py:42]: Instead of piping types use `Optional[list]`

File: autogpt/config/ai_config.py [hunk 3]
```diff
@@ -46,71 +49,157 @@ def __init__(
...
         """
         if ai_goals is None:
             ai_goals = []
+        if plugins is None:
+            plugins = []
         self.ai_name = ai_name
         self.ai_role = ai_role
         self.ai_goals = ai_goals
         self.api_budget = api_budget
+        self.plugins = plugins
```
Comment (erik-megarad) [inline autogpt/config/ai_config.py:64]:
```suggestion
        self.plugins = plugins or []
```

This way you don't need lines 58-59

File: autogpt/config/ai_config.py [hunk 4]
```diff
@@ -46,71 +49,157 @@ def __init__(
...
-        # type: Type[AIConfig]
-        return AIConfig(ai_name, ai_role, ai_goals, api_budget)
-
-    def save(self, config_file: str = SAVE_FILE) -> None:
+            all_configs = {}
+
+        configs = all_configs.get("configs", {}) or {}
+
+        ai_configs = {}
+        for ai_name, config_params in configs.items():
```
Comment (erik-megarad) [inline autogpt/config/ai_config.py:99]: Would this be a breaking change with existing `ai_settings.yaml` files? If possible a backwards-compatible change would be good. We've already had so many breaking changes lately.

File: autogpt/prompts/prompt.py [hunk 5]
```diff
@@ -46,13 +46,304 @@ def build_default_prompt_generator() -> PromptGenerator:
     return prompt_generator
 
 
-def construct_main_ai_config() -> AIConfig:
+def construct_main_ai_config(config_file: str = None) -> AIConfig:
```
Comment (erik-megarad) [inline autogpt/prompts/prompt.py:49]: This is an extremely long function. Would it be possible to break it up? e.g. the functionality for creating, changing, and deleting configs could be separated cleanly.

File: autogpt/setup.py [hunk 6]
```diff
@@ -74,97 +75,260 @@ def prompt_user() -> AIConfig:
...
+                f"Enter up to {max_goals} goals for your AI:",
+                Fore.GREEN,
+                "use [Enter] to keep the current goal / save the input.",
+                speak_text=True,
+            )
+        ai_goals = list(default_goals)
+        kept_goals = []
+        deleted_goals_indices = []
+
+        for i in range(len(ai_goals)):
```
Comment (erik-megarad) [inline autogpt/setup.py:182]: Seems easier to just do

```suggestion
        for ai_goal in ai_goals:
```

File: autogpt/prompts/prompt.py [hunk 7]
```diff
@@ -46,13 +46,304 @@ def build_default_prompt_generator() -> PromptGenerator:
...
+                logger.typewriter_log(f"{i}. {cfg_name}")
+            logger.typewriter_log(f"{len(all_configs) + 1}. Create new configuration")
+            logger.typewriter_log(
+                f"{len(all_configs) + 2}. Change existing configuration"
+            )
+            logger.typewriter_log(
+                f"{len(all_configs) + 3}. Delete existing configuration"
+            )
+            selection = clean_input("Please choose a number: ")
+            if selection.isdigit() and 0 < int(selection) <= len(all_configs) + 3:
```
Comment (erik-megarad) [inline autogpt/prompts/prompt.py:90]: It doesn't feel great to be so reliant on list indexing / `len`. I think adding an intermediate set of variables or data structure indicating the user's selection would be really nice. That way the logic below reads cleaner, and if we decide to change the way that the menu works we wouldn't need to change all of the code.

**Review by erik-megarad — CHANGES_REQUESTED** [review]

File: autogpt/prompts/prompt.py (hunk 5 above)
Comment (bassie661) [reply autogpt/prompts/prompt.py:49]: Thanks and absolutely right, I was already busy with it. Can you hook up on discord, I would like to hear your opinion, maybe in the end it's even better to close this one and re-factor on a new PR.

File: autogpt/prompts/prompt.py (hunk 7 above)
Comment (bassie661) [reply autogpt/prompts/prompt.py:90]: True it should also be DRY and easier to maintain, will look at it

File: autogpt/config/ai_config.py (hunk 1 above)
Comment (bassie661) [reply autogpt/config/ai_config.py:34]: The whole SAVE_FILE approach is weird in my eyes, I am looking into setting the cfg.ai_settings_filepath in main.py so it just has to be set once. it's fixed anyway.

**Comment by bassie661** [comment]
closing this PR, because of the review I decided to write a complete different approach on a new branch, see #4646

**Pull request closed without merging**